### PR TITLE
Add src_dims to reshape operator

### PR DESCRIPTION
**Pull request by majra20**
Signed-off-by: Rafal Maj <rmaj@nvidia.com>

#### Why we need this PR?
- It adds new feature needed because of we want to add squeeze and expandsdim operators and src_dims will allow us to implement them.

#### What happened in this PR?
 - What solution was applied:
     - Added src_dims to reshape operator
 - Affected modules and functionalities:
     - Reshape operator allows to use src_dims to specify which dimensions and in which order should be kept
 - Key points relevant for the review:
     - Remove only dims with shape 1
     - New dims are added when -1 value provided
     - Shapes are reordered
 - Validation and testing:
     - Python tests
 - Documentation (including examples):
     - NA

Dali-1913

### CODE REVIEW

**Review by jantonguirao — APPROVED** [review]

File: dali/operators/generic/reshape.cc [hunk 1]
```diff
@@ -63,7 +63,20 @@ If a value is not specified, if number of dimension matches existing layout, the
 layout is preserved. If the number of dimensions does not match, the argument is reset
 to empty. If a value is set, and is not empty, the layout must match the dimensionality
 of the output.)code",
-                  TensorLayout(""));
+                  TensorLayout(""))
+  .AddOptionalArg("src_dims", R"code(Indices of dimensions to keep.
+
+This argument can be used to manipulate the order of existing dimensions or to remove or
+add dimensions with extent one. A special index value -1 can be used to insert new dimensions
+at any particular position.
```
Comment (mzient) [inline dali/operators/generic/reshape.cc:71]:
```suggestion
This argument can be used to manipulate the order of existing dimensions or to remove or
add dimension. A special index value -1 can be used to insert new dimensions.
```

File: dali/operators/generic/reshape.cc [hunk 2]
```diff
@@ -63,7 +63,20 @@ If a value is not specified, if number of dimension matches existing layout, the
...
+
+For example, reshaping a sample with shape ``[300, 200, 1]`` and a ``src_dims``
+argument ``[-1, 1, 0]`` produces an output shape ``[1, 200, 300]``. A leading dimension with
+extent 1 is inserted at the beginning, followed by the first original dimensions but in reverse
+order. The last dimension is removed.
+
+All the indices should be in the range of valid dimensions of the input, or -1.)code",
```
Comment (mzient) [inline dali/operators/generic/reshape.cc:78]:
```suggestion
All indices must be in the range of valid dimensions of the input, or -1.)code",
```

File: dali/operators/generic/reshape.cc [hunk 3]
```diff
@@ -63,7 +63,20 @@ If a value is not specified, if number of dimension matches existing layout, the
...
+
+This argument can be used to manipulate the order of existing dimensions or to remove or
+add dimensions with extent one. A special index value -1 can be used to insert new dimensions
+at any particular position.
+
+For example, reshaping a sample with shape ``[300, 200, 1]`` and a ``src_dims``
+argument ``[-1, 1, 0]`` produces an output shape ``[1, 200, 300]``. A leading dimension with
+extent 1 is inserted at the beginning, followed by the first original dimensions but in reverse
+order. The last dimension is removed.
+
```
Comment (mzient) [inline dali/operators/generic/reshape.cc:77]: I think we were to support reindexing with `rel_shape`, too? I think it deserves a mention.
```
The `src_dims` argument can be used together with `rel_shape`, in which case the relative extents in `rel_shape` describe to the target dimensions. In the example above, specifying ``rel_shape = [-1, 0.5, 2]`` would result in the output shape ``[1, 100, 600]``.
```

File: dali/operators/generic/reshape.cc [hunk 4]
```diff
@@ -89,13 +102,20 @@ Reshape<Backend>::Reshape(const OpSpec &spec) : Base(spec) {
   bool has_shape_arg = spec.HasArgument("shape") || spec.HasTensorArgument("shape");
   bool has_layout_arg = spec.HasArgument("layout");
   bool has_rel_shape_arg = spec.HasArgument("rel_shape") || spec.HasTensorArgument("rel_shape");
+  bool has_src_dims_arg = spec.HasArgument("src_dims");
+  bool has_axes_arg = spec.HasArgument("axes");
+  bool has_axis_names_arg = spec.HasArgument("axis_names");
```
Comment (mzient) [inline dali/operators/generic/reshape.cc:107]: These should not be a part of Reshape.

File: dali/operators/generic/reshape.cc [hunk 5]
```diff
@@ -89,13 +102,20 @@ Reshape<Backend>::Reshape(const OpSpec &spec) : Base(spec) {
...
+  }
   if (spec.HasArgument("dtype"))
     output_type_id_ = spec.GetArgument<DALIDataType>("dtype");
   DALI_ENFORCE(has_shape_input + has_shape_arg + has_rel_shape_arg <= 1, make_string(OpName(),
     ": shape input, `shape` argument and `rel_shape` argument are mutually exclusive"));
 
-  bool does_reshape = has_shape_input || has_shape_arg || has_rel_shape_arg || has_layout_arg;
-  if (!has_shape_input && !has_shape_arg && !has_rel_shape_arg && !has_layout_arg) {
+  if (!has_shape_input && !has_shape_arg && !has_rel_shape_arg && !has_layout_arg
+      && !has_src_dims_arg && !has_axes_arg && !has_axis_names_arg) {
```
Comment (mzient) [inline dali/operators/generic/reshape.cc:118]:
```suggestion
      && !has_src_dims_arg) {
```

File: dali/operators/generic/reshape.cc [hunk 6]
```diff
@@ -247,10 +265,19 @@ void Reshape<Backend>::CalculateOutputShape(const Workspace &ws) {
   switch (shape_source_) {
     case ShapeSource::Arg:
       if (use_rel_shape_) {
+        if (!src_dims_.empty()) {
+          DALI_ENFORCE(rel_uniform_shape_.size() == src_dims_.size(),
+            make_string(OpName(), ": ``src_dims`` and ``rel_shape`` should have the same"
+            " length when both of them are provided. Got ", src_dims_.size(), " and ",
+            rel_uniform_shape_.size(), " elements respectively"));
```
Comment (mzient) [inline dali/operators/generic/reshape.cc:272]: Not a must - just a suggestion. I think we can make it shorter ans skip the part about "when both are provided", becuase they obviously were, if we hit this error:
```suggestion
            make_string(OpName(), ": ``src_dims`` and ``rel_shape`` have different"
            " lengths: ", src_dims_.size(), " vs ", rel_uniform_shape_.size()));
```

File: dali/operators/generic/reshape.cc [hunk 7]
```diff
@@ -370,6 +413,23 @@ void Reshape<GPUBackend>::RunImpl(DeviceWorkspace &ws) {
...
+
+  const auto &in = ws.template InputRef<Backend>(0);
+  const auto &input_shape = in.shape();
+  const int ndim = input_shape.sample_dim();
+  for (size_t d = 0; d < src_dims_.size(); d++) {
+    DALI_ENFORCE(-1 <= src_dims_[d] && src_dims_[d] < ndim,
+      make_string(OpName(), ": Out of bounds ``src_dims`` index. The indices in ``src_dims``"
+      " should be either a valid dimension index (range 0..ndim-1) or -1 to insert a new dimension."
+      " Got: src_dims[", d, "]=", src_dims_[d], ", ndim=", ndim));
+  }
```
Comment (mzient) [inline dali/operators/generic/reshape.cc:430]:
```suggestion
        make_string(OpName(), ": ``src_dims[", i, "]`` == ", src_dims[i], " is out of bounds.\n"
        "The indices in ``src_dims`` should be either valid dimension indices in range [0..", ndim-1, "] "
        "or -1 to insert a new dimension.")
  }
```
There's also the case of 0D input (in which case the only valid src_dim is -1), but I don't think it deserves a dedicated message.

File: dali/operators/generic/reshape.h [hunk 8]
```diff
@@ -41,30 +42,40 @@ class Reshape : public Operator<Backend> {
 
   void RunImpl(Workspace &ws) override;
 
+ protected:
+  virtual void CalculateOutputShape(const Workspace &ws);
```
Comment (mzient) [inline dali/operators/generic/reshape.h:46]:
```suggestion
  virtual void CalculateOutputShape(const Workspace &ws);
  
```

File: dali/operators/generic/reshape.h [hunk 9]
```diff
@@ -41,30 +42,40 @@ class Reshape : public Operator<Backend> {
...
+  void CheckSrcDims(const Workspace &ws);
+  inline void SetOutputType(const Workspace &ws) {
```
Comment (mzient) [inline dali/operators/generic/reshape.h:48]:
```suggestion
  virtual void CalculateOutputShape(const Workspace &ws);
  
  void CheckSrcDims(const Workspace &ws);
  
  inline void SetOutputType(const Workspace &ws) {
```

File: dali/operators/generic/reshape.cc [hunk 10]
```diff
@@ -247,10 +266,18 @@ void Reshape<Backend>::CalculateOutputShape(const Workspace &ws) {
   switch (shape_source_) {
     case ShapeSource::Arg:
       if (use_rel_shape_) {
+        if (!src_dims_.empty()) {
```
Comment (mzient) [inline dali/operators/generic/reshape.cc:269]: This fails to capture the case of dropping _all_ dimesnions from a scalar value.

File: dali/operators/generic/reshape.cc [hunk 11]
```diff
@@ -270,7 +297,23 @@ void Reshape<Backend>::CalculateOutputShape(const Workspace &ws) {
       ShapeFromInput(ws.template InputRef<CPUBackend>(1), false);
       break;
     case ShapeSource::None:
-      output_shape_ = input_shape_;
+      if (src_dims_.empty()) {
```
Comment (mzient) [inline dali/operators/generic/reshape.cc:300]: likewise

File: dali/operators/generic/reshape.cc [hunk 12]
```diff
@@ -370,6 +413,23 @@ void Reshape<GPUBackend>::RunImpl(DeviceWorkspace &ws) {
   out.SetLayout(layout);
 }
 
+template <typename Backend>
+void Reshape<Backend>::CheckSrcDims(const Workspace &ws) {
+  if (src_dims_.empty()) {
```
Comment (mzient) [inline dali/operators/generic/reshape.cc:418]: and again

**Review by mzient — APPROVED** [review]

**Comment by mzient** [comment]
!build

**Pull request merged**